File: cpp/_1.cpp
/**
1. Two Sum

Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

Example:

Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].

*/
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> m;
        vector<int> result;
        for(int i=0; i<nums.size(); i++){
        // not found the second one
            if (m.find(nums[i])==m.end() ) {
            // store the first one poisition into the second one's key
                m[target - nums[i]] = i;
            }else {
            // found the second one
                result.push_back(m[nums[i]]);
                result.push_back(i);
                break;
            }
        }
        return result;
    }
};

File: javascript/_1.js
// Two sum

var Solution1 = function (nums, target) {
  var sum = {}
  for (var i = 0; i < nums.length; i++) {
    if (!Object.prototype.hasOwnProperty.call(sum, (target - nums[i]))) {
      sum[nums[i]] = i
    } else {
      return [i, sum[target - nums[i]]]
    }
  }
}

var Solution2 = function (nums, target) {
  var  p1 = 0
  var  p2 = nums.length - 1
  while (p1 < p2){
    if (nums[p1] + nums[p2] < target) {
      p1 += 1
    }
    else if (nums[p1] + nums[p2] > target) {
      p2 -= 1
    }
    else {
      return [ p1, p2]  
    }
  }
}

console.log(Solution1([2, 7, 11, 15], 9))
console.log(Solution2([2, 7, 11, 15], 9))
